internal_ws/ykcompile/src/lib.rs

```diff
@@ -1368,24 +1368,22 @@ impl TraceCompiler {
         // Reserved memory on the stack to spill live locals to during a guard failure.
         let live_off = REG_POOL.len() * usize::try_from(QWORD_REG_SIZE).unwrap();
 
-        // FIXME Can we re-review this whole comment, as I'm sure it could be made shorter, and I'm
-        // not sure it's right to say that the stack must be 16-byte aligned *before* the call. I
-        // thought it was right *after* the call (once the RV has been pushed).
-        //
         // As we'll be making calls to other functions using the Sys-V 64 ABI, we need to make sure
-        // the stack is aligned to 16 bytes before the call instruction. Note that the call
-        // instruction itself will push the return address onto the stack, so at the beginning of
-        // the trace the stack will have an extra offset of 8 bytes.
+        // the stack is correctly aligned. We must ensure that the stack pointer is 16-byte aligned
+        // immediately before the call instruction is executed.
+        //
+        // FIXME The following is not going to work in the long run, as we will have to push to the
```

Comment:
Do you mean calls we do in `c_call`? The problem with `c_call` is that it calls Rust functions, and there is no nice way to wrap those into sys-v calls, so we've agreed to use the unofficial/unstable Rust ABI for those, so I don't think we need to worry about this. However, we can assume that when we enter `c_call`, the stack is 16-bytes aligned, since it's aligned at the beginning of the trace, and we don't misalign it anywhere (i.e. if we need to push to the stack we always pop immediately after before compiling further instructions).

Reply:
I would assume that c_call could call anything, via any ABI (although obviously its implementation is incomplete).

It wouldn't surprise me, for example, if there is some sysv function in the stdlib (that we can't trace).

Reply:
Fair point. But this doesn't change the fact that we will be aligned when we enter the `c_call` function. So if the call requires arguments to be pushed onto the stack, then we just need to count them and make sure we align again before compiling the `call` instruction.

Reply:
Good point. I think the wording just needs to be tweaked somehow. Perhaps you can propose something?